cmd/kubeadm/app/cmd/join.go

```diff
@@ -235,34 +279,228 @@ func (j *Join) Validate(cmd *cobra.Command) error {
+
+		// checks if the certificates for the control plane were already provided by the user in
+		// the local pki folder and checks if those certs are compliant with the requirements
+		// ** Note for the reviewers:
+		// ** see notes in phases/cert about how sharing validate cert functions between the two features
+		err := certsphase.CheckIfLocalPkiReadyForJoinAsMaster(masterConfiguration)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// PrepareForJoinAsMaster makes all preparation activities require for a node joining as a master
+func (j *Join) PrepareForJoinAsMaster(masterConfiguration *kubeadmapi.MasterConfiguration) error {
+
+	// if the certificates for the control plane are stored in a local pki folder
+	// (if certificates are not stored in secrets)
+	if !features.Enabled(masterConfiguration.FeatureGates, features.StoreCertsInSecrets) {
+
+		// Creates the missing certificates
+		// NB. This function will use the certificate authority and service account signing key
+		//     provided by the user before executing init --master.
+		//     (those certificates/keys must be the same across all master instances)
+		certsphase.CreatePKIAssets(masterConfiguration)
+	}
+
+	// if the cluster uses StaticPods manifests (if the cluster is not self hosted)
+	if !features.Enabled(masterConfiguration.FeatureGates, features.SelfHosting) {
+
+		// Generate kubeconfig files for controller manager, scheduler and for the admin/kubeadm itself
+		// NB. The kubeconfig file for kubelet will be generated by the TLS bootstrap process in
+		// following steps of the join --master workflow
+		err := kubeconfigphase.CreateJoinMasterKubeConfigFiles(kubeadmconstants.KubernetesDir, masterConfiguration)
+		if err != nil {
+			return fmt.Errorf("error generating kubeconfig files: %v", err)
+		}
+
+		// Creates static pod manifests file for the control plane components to be deployed on this node
+		// Static pods will be created and managed by the kubelet as soon as it starts
+		err = controlplanephase.CreateInitStaticPodManifestFiles(kubeadmconstants.GetStaticPodDirectory(), masterConfiguration)
+		if err != nil {
+			return fmt.Errorf("error creating static pod manifest files for the control plane components: %v", err)
+		}
+
+		return nil
+	}
+
+	// Otherwise, if the cluster uses a SelfHosting control plane
+
+	// Creates the admin kubeconfig file for the admin and for kubeadm itself.
+	err := kubeconfigphase.CreateAdminKubeConfigFile(kubeadmconstants.KubernetesDir, masterConfiguration)
+	if err != nil {
+		return fmt.Errorf("error generating the admin kubeconfig file: %v", err)
+	}
+
+	// NB. self hosted clusters desn't requires preparation activities,
+	// because the existing DeamonSets for the control plane components will automatically trigger
+	// the deploymento of new instances of such components on this node as soon as will join the cluster.
+
+	return nil
```

Comment:
we should add more logging so that users know what's happening (e.g. "Creating static pods")

Reply:
Basically we are calling the same methods used in `kubeadm init`, so we already get the same level of logging also here. However I will planning to give a final pass on logging/ux as soon as the PR stabilizes